### PR TITLE
own: add API for assigning an owner.

### PR DESCRIPTION
**Pull request by sashaostrikov**
Test plan:
GraphQL tests added.

The first part of https://github.com/sourcegraph/sourcegraph/issues/52135

### CODE REVIEW

File: cmd/frontend/graphqlbackend/own.go [hunk 1]
```diff
@@ -137,6 +140,17 @@ type DeleteCodeownersFilesInput struct {
 	RepoName *string
 }
 
+type AssignOwnerArgs struct {
+	Input AssignOwnerInput
+}
+
+type AssignOwnerInput struct {
+	AssignedOwnerID   graphql.ID
```
Comment (cbart) [inline cmd/frontend/graphqlbackend/own.go:148]: That's user ID, right?

```suggestion
	AssignedOwnerID   graphql.ID  // user ID
```

File: cmd/frontend/graphqlbackend/own.go [hunk 2]
```diff
@@ -137,6 +140,17 @@ type DeleteCodeownersFilesInput struct {
...
+	RepoID            graphql.ID
+	AbsolutePath      string
+	WhoAssignedUserID graphql.ID
```
Comment (cbart) [inline cmd/frontend/graphqlbackend/own.go:151]: Should we have a parameter here? Can that be taken from the context?

File: enterprise/cmd/frontend/internal/own/resolvers/resolvers.go [hunk 3]
```diff
@@ -770,3 +771,55 @@ func userifyPatterns(patterns []string) (results []string) {
...
+
+func (r *ownResolver) AssignOwner(ctx context.Context, args *graphqlbackend.AssignOwnerArgs) (*graphqlbackend.EmptyResponse, error) {
+	if err := rbac.CheckCurrentUserHasPermission(ctx, r.db, rbac.OwnershipAssignPermission); err != nil {
+		return nil, err
+	}
+	u, err := unmarshalAssignOwnerArgs(args.Input)
+	if err != nil {
+		return nil, err
+	}
+	err = r.db.AssignedOwners().Insert(ctx, u.AssignedOwnerID, u.RepoID, u.AbsolutePath, u.WhoAssignedUserID)
```
Comment (cbart) [inline enterprise/cmd/frontend/internal/own/resolvers/resolvers.go:783]: (similar comment to the API one) What do we do if `WhoAssignedUserID` is different from the user that we're checking permissions for?

File: enterprise/cmd/frontend/internal/own/resolvers/resolvers_test.go [hunk 4]
```diff
@@ -1393,9 +1393,101 @@ func TestOwnership_WithAssignedOwners(t *testing.T) {
...
+			Message: "user is missing permission OWNERSHIP#ASSIGN",
+			Path:    []any{"assignOwner"},
+		}}
+		baseTest.ExpectedErrors = expectedErrs
+		baseTest.ExpectedResult = `{"assignOwner":null}`
+		graphqlbackend.RunTest(t, baseTest)
+	})
+
+	t.Run("bad request", func(t *testing.T) {
+		baseTest.Context = adminCtx
```
Comment (cbart) [inline enterprise/cmd/frontend/internal/own/resolvers/resolvers_test.go:1472]: you probably need to deep copy basetest here, as state will leak depending on test order right? For instance in this test you're assigning `Variables` but in the test above they're not re-assigned or emptied.

Alternative baseTest can just be turned into a helper function and invoked in every test - then data is isolated.

**Review by cbart — APPROVED** [review]
:shipit:

File: cmd/frontend/graphqlbackend/own.go (hunk 1 above)
Comment (sashaostrikov) [reply cmd/frontend/graphqlbackend/own.go:148]: yep

File: cmd/frontend/graphqlbackend/own.go (hunk 2 above)
Comment (sashaostrikov) [reply cmd/frontend/graphqlbackend/own.go:151]: that's right, I'll take it from context!

File: enterprise/cmd/frontend/internal/own/resolvers/resolvers.go (hunk 3 above)
Comment (sashaostrikov) [reply enterprise/cmd/frontend/internal/own/resolvers/resolvers.go:783]: will use the ID from context

File: enterprise/cmd/frontend/internal/own/resolvers/resolvers_test.go (hunk 4 above)
Comment (sashaostrikov) [reply enterprise/cmd/frontend/internal/own/resolvers/resolvers_test.go:1472]: that's a good point, I'll fix it

**Pull request merged**